sound/soc/soc-pcm.c

```diff
@@ -1048,20 +1048,44 @@ static int __soc_pcm_hw_params(struct snd_soc_pcm_runtime *rtd,
 	}
 
 	for_each_rtd_cpu_dais(rtd, i, cpu_dai) {
+		struct snd_pcm_hw_params cpu_params;
+		unsigned int ch_mask = 0;
+		int j;
+
 		/*
 		 * Skip CPUs which don't support the current stream
 		 * type. See soc_pcm_init_runtime_hw() for more details
 		 */
 		if (!snd_soc_dai_stream_valid(cpu_dai, substream->stream))
 			continue;
 
-		ret = snd_soc_dai_hw_params(cpu_dai, substream, params);
+		/* copy params for each cpu */
+		cpu_params = *params;
+
+		if (!rtd->dai_link->codec_ch_maps)
+			goto hw_params;
+		/*
+		 * construct cpu channel mask by combining ch_mask of each
+		 * codec which maps to the cpu.
+		 */
+		for_each_rtd_codec_dais(rtd, j, codec_dai) {
+			if (rtd->dai_link->codec_ch_maps[j].connected_cpu_id != i)
+				continue;
+			ch_mask |= rtd->dai_link->codec_ch_maps[j].ch_mask;
```

Comment:
do we need to check M < N?

Reply:
still wondering if this needs a test, or it's assumed that the test was already checked?

Reply:
> still wondering if this needs a test, or it's assumed that the test was already checked?

M < N was checked in snd_soc_dapm_connect_dai_link_widgets(). We provided a log to show num_cpus and/or num_codecs may be wrong. Here, we get ch_mask for each cpu. We will only calculate the ch_mask when connected_cpu_id is equal to the cpu dai id, and if the cpu is not connected to any codec, ch_mask will be 0. So, I think it should be fine.